Aliquot Sum, Narcissistic Number (Armstrong Number), Perfect Cube, Perfect Square, Perfect Number added

simolab99 opened the pull request:
Aliquot Sum, Narcissistic Number (Armstrong Number), Perfect Cube, Perfect Square, Perfect Number added.

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have made corresponding changes to the comments
- [ ] I have made corresponding changes to the README.md

siriak commented on Algorithms.Tests/Numeric/NarcissisticNumberTest.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public static class NarcissisticNumberTest
+    {
+        [Test]
+        [TestCase(2)]
+        [TestCase(3)]
+        [TestCase(153)]
+        [TestCase(371)]
+        public static void NarcissisticNumberWork(int number)
+        {
+            // Arrange
+
+            // Act
+            var result = NarcissisticNumber.IsNarcissistic(number);
+
+            // Assert
+            Assert.IsTrue(result);
+        }
```
Please add tests to check numbers that are not narcissistic

siriak commented on Algorithms.Tests/Numeric/PerfectCubeTests.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public static class PerfectCubeTests
+    {
+        [Test]
+        [TestCase(8)]
+        [TestCase(27)]
+        [TestCase(64)]
+        [TestCase(216)]
+        public static void PerfectCubeWork(int number)
+        {
+            // Arrange
+
+            // Act
+            var result = PerfectCube.IsPerfectCube(number);
+            
+            // Assert
+            Assert.IsTrue(result);
+        }
```
The same here, add negative cases

siriak commented on Algorithms.Tests/Numeric/PerfectNumberTest.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public static class PerfectNumberTests
+    {
+        [Test]
+        [TestCase(6)]
+        [TestCase(28)]
+        [TestCase(496)]
+        [TestCase(8128)]
+        public static void PerfectNumberWork(int number)
+        {
+            // Arrange
+
+            // Act
+            var result = PerfectNumber.IsPerfectNumber(number);
+
+            // Assert
+            Assert.IsTrue(result);
+        }
```
And here

siriak commented on Algorithms.Tests/Numeric/PerfectSquareTest.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public static class PerfectSquareTests
+    {
+        [Test]
+        [TestCase(4)]
+        [TestCase(9)]
+        [TestCase(16)]
+        [TestCase(81)]
+        public static void PerfectSquareWork(int number)
+        {
+            // Arrange
+
+            // Act
+            var result = PerfectSquare.IsPerfectSquare(number);
+
+            // Assert
+            Assert.IsTrue(result);
+        }
```
And here

siriak commented on Algorithms/Numeric/Aliquot.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    /// <summary>
+    /// In number theory, the aliquot sum s(n) of a positive integer n is the sum of all proper divisors
+    /// of n, that is, all divisors of n other than n itself.For example, the proper divisors of 15
```
```suggestion
    /// of n, that is, all divisors of n other than n itself. For example, the proper divisors of 15

```

siriak commented on Algorithms/Numeric/NarcissisticNumber.cs:
```diff
@@ -0,0 +1,43 @@
+        /// <summary>
+        /// Checks if a number is a Narcissistic number or not.
+        /// </summary>
+        /// <param name="number">Number too check.</param>
```
```suggestion
        /// <param name="number">Number to check.</param>

```

siriak commented on Algorithms/Numeric/PerfectCube.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <summary>
+        /// Checks if a number is a perfect cube or not.
+        /// </summary>
+        /// <param name="number">Number too check.</param>
```
```suggestion
        /// <param name="number">Number to check.</param>

```

siriak commented on Algorithms/Numeric/PerfectCube.cs:
```diff
@@ -0,0 +1,27 @@
+        /// </summary>
+        /// <param name="number">Number too check.</param>
+        /// <returns>True if is a perfect cube; False otherwise.</returns>
+        /// <exception cref="ArgumentException">Error number is not on interval (0.0; int.MaxValue).</exception>
+        public static bool IsPerfectCube(int number)
+        {
+            if (number < 0)
+            {
+                throw new ArgumentException(string.Format("Error number is not on interval (0.0; {0}).", int.MaxValue));
+            }
+
```
```suggestion
        public static bool IsPerfectCube(int number)
        {
```

siriak replied:
Perfect cubes can be negative AFAIK

siriak commented on Algorithms/Numeric/PerfectNumber.cs:
```diff
@@ -0,0 +1,37 @@
+        /// <summary>
+        /// Checks if a number is a perfect number or not.
+        /// </summary>
+        /// <param name="number">Number too check.</param>
+        /// <returns>True if is a perfect number; False otherwise.</returns>
+        /// <exception cref="ArgumentException">Error number is not on interval (0.0; int.MaxValue).</exception>
+        public static bool IsPerfectNumber(int number)
+        {
+            if (number < 0)
+            {
+                throw new ArgumentException(string.Format("Error number is not on interval (0.0; {0}).", int.MaxValue));
```
```suggestion
        /// <param name="number">Number too check.</param>
        /// <returns>True if is a perfect number; False otherwise.</returns>
        /// <exception cref="ArgumentException">Error number is not on interval (0.0; int.MaxValue).</exception>
        public static bool IsPerfectNumber(int number)
        {
            if (number < 0)
            {
                throw new ArgumentException($"{nameof(number)} cannot be negative");

```

siriak commented on Algorithms/Numeric/PerfectSquare.cs:
```diff
@@ -0,0 +1,27 @@
+        {
+            if (number < 0)
+            {
+                throw new ArgumentException(string.Format("Error number is not on interval (0.0; {0}).", int.MaxValue));
```
```suggestion
                return false;

```

siriak reviewed (commented):
Hi, looks good. Could you add some more tests and address some comments I left? I cannot merge unless all the checks pass.

simolab99 commented:
I corrected all the above and deleted PerfectCube.cs and PerfectCubeTest.cs because I'm not sure how to get it working with the current Math.Pow() without resulting in a NaN

siriak commented on Algorithms/Numeric/Aliquot.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            if (number < 0)
+            {
+                throw new ArgumentException(string.Format("Error number is not on interval (0.0; {0}).", int.MaxValue));
```
```suggestion
                throw new ArgumentException(string.Format($"{nameof(number)} cannot be negative"));

```

siriak reviewed (approved):
Looks good, thanks!

siriak commented on Algorithms/Numeric/Aliquot.cs:
```diff
@@ -0,0 +1,39 @@
+            if (number < 0)
+            {
+                throw new ArgumentException($"{nameof(number)} cannot be negative");
+
```
```suggestion

```

siriak commented on Algorithms/Numeric/NarcissisticNumber.cs:
```diff
@@ -0,0 +1,43 @@
+            while (number > 0)
+            {
+                int remainder = number % 10;
+                int power = 1;
+                for (int i = 1; i <= numberOfDigits; power *= remainder, ++i)
+                {
+                }
```
```suggestion
                int power = (int)Math.Pow(remainder, numberOfDigits);

```

siriak reviewed (approved):
Could you add some tests to cover input of negative numbers for Aliquot sum and perfect numbers? Test coverage is too low, cannot merge until checks pass.

simolab99 commented:
Now all the checks pass but I don't understand why codecov/patch throws a warning on a covered line.

siriak reviewed (approved):
Looks good, thanks!
